### PR TITLE
add boilerplate code; change python version

### LINKED ISSUE #11: set up job quality repo + bucket + cookiecutter
do it in open jobs lake (bucket)!!

### PR DESCRIPTION
**Pull request by india-kerle**
---

# Description

This PR:
- adds getters to load an old sample of OJO data incl. descriptions we can use for the labelling exercise;
- adds general data getters to load and save data to s3;
- updates the python version in `environment.yaml`
- Adds additional utils to embed texts and chunk objects

Fixes # (issue)

This should close #11 

# Instructions for Reviewer

I haven't added any tests but you can set up the environment (`make install`) and make sure the getters work in a jupyter notebook. 

i.e.
```
from dap_job_quality.getters.ojo_getters import get_ojo_sample

data = get_ojo_sample()
```

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

File: environment.yaml [hunk 1]
```diff
@@ -3,6 +3,6 @@ channels:
   - conda-forge
 dependencies:
   - pip
-  - python=3.8
+  - python=3.10
```
Comment (RFOxbury) [inline environment.yaml:6]: I checked out the branch and ran `make install` and now when I activate `dap_job_quality` and run `python --version` it says I'm on python 3.11.4... Don't know if this is a problem or not but it seemed surprising

File: dap_job_quality/utils/bert_vectorizer.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    embeddings = np.concatenate(embeddings)
+
+    if not id_list:
+        id_list = sent_list
+
+    # create dict
+    return dict(zip(id_list, embeddings))
+
+
+class BertVectorizer:
```
Comment (RFOxbury) [inline dap_job_quality/utils/bert_vectorizer.py:41]: Looks really nice!

Comment (RFOxbury) [reply dap_job_quality/utils/bert_vectorizer.py:41]: I wish I could write a Bert vectorizer that quickly

File: requirements.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+pandas
+boto3
+pyarrow
+fsspec
+s3fs
```
Comment (RFOxbury) [inline requirements.txt]: FYI when I load `ojo_getters` I get this warning:
```
UserWarning: Your installed version of s3fs is very old and known to cause
severe performance issues, see also https://github.com/dask/dask/issues/10276
```
... Is it fine to ignore this?

File: dap_job_quality/getters/ojo_getters.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import pandas as pd
+
+
+# Let's currently get a sample of the data from PRINZ to work with
+# for labelling etc.
+def get_ojo_sample() -> pd.DataFrame:
+    """Gets ojo sample data from s3
+
+    Returns:
+        pd.Dataframe: ojo sample data
```
Comment (RFOxbury) [inline dap_job_quality/getters/ojo_getters.py]:
```suggestion
        pd.Dataframe: ojo sample data with the fields:
            - `id`: unique identifier for the job ad
            - `job_title_raw`
            - `created`: datetime when the job was listed?
            - `description`: job description
            - `itl_3_code`
            - `itl_3_name`
```

File: dap_job_quality/getters/ojo_getters.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+import os
+from typing import Dict, List
+import pandas as pd
+
+
+# Let's currently get a sample of the data from PRINZ to work with
+# for labelling etc.
+def get_ojo_sample() -> pd.DataFrame:
+    """Gets ojo sample data from s3
```
Comment (RFOxbury) [inline dap_job_quality/getters/ojo_getters.py]:
```suggestion
    """Gets ojo sample data (100,000 job ads) from s3.
```

File: dap_job_quality/getters/ojo_getters.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        PRINZ_BUCKET_NAME,
+        "outputs/data/ojo_application/deduplicated_sample/ojo_sample.csv",
+    )
+
+
+def get_ojo_job_title_sample() -> pd.DataFrame:
+    """Gets ojo sample data with job title and sectors information from s3
+
+    Returns:
+        pd.Dataframe: ojo sample data
```
Comment (RFOxbury) [inline dap_job_quality/getters/ojo_getters.py]:
```suggestion
        pd.Dataframe: ojo sample data with the following fields:
            - id: unique identifier for the job ad
            - company_raw: company name
            - job_title_raw
            - job_location_raw
            - created: datetime when the job was listed?
            - type: who listed the ad (e.g. "Recruitement consultancy", "Direct employer")
            - sector
            - parent_sector
            - knowledge_domain: eg "Engineering", "Legal"
            - occupation: standardised job title eg "Manager consultant"
```

File: dap_job_quality/getters/ojo_getters.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        PRINZ_BUCKET_NAME,
+        "outputs/data/ojo_application/deduplicated_sample/job_title_data_sample.csv",
+    )
+
+
+def get_ojo_location_sample() -> pd.DataFrame:
+    """Gets ojo sample data with location information from s3
+
+    Returns:
+        pd.Dataframe: ojo sample data
```
Comment (RFOxbury) [inline dap_job_quality/getters/ojo_getters.py]:
```suggestion
        pd.Dataframe: ojo sample data with the fields:
            - id: unique identifier of the job ad
            - job_location_raw
            - itl_1_code: eg "TLJ"
            - itl_1_name: eg "South East (England)
            - itl_2_code
            - itl_2_name
            - itl_3_code
            - itl_3_name
            - is_uk: 1.0 if true, 0.0 if false
            - is_large_geo: 1.0 if true, 0.0 if false (what does this mean?)
            - location: standardised from job_location_raw
            - coordinates
```

File: dap_job_quality/getters/ojo_getters.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        PRINZ_BUCKET_NAME,
+        "outputs/data/ojo_application/deduplicated_sample/locations_data_sample.csv",
+    )
+
+
+def get_ojo_salaries_sample() -> pd.DataFrame:
+    """Gets ojo sample data with salaries information from s3
+
+    Returns:
+        pd.Dataframe: ojo sample data
```
Comment (RFOxbury) [inline dap_job_quality/getters/ojo_getters.py]:
```suggestion
        pd.Dataframe: ojo sample data with the fields:
            - id: unique identifier of the job ad
            - min_annualised_salary
            - max_annualised_salary
```

File: dap_job_quality/getters/ojo_getters.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        PRINZ_BUCKET_NAME,
+        "outputs/data/ojo_application/deduplicated_sample/salaries_data_sample.csv",
+    )
+
+
+def get_ojo_skills_sample() -> pd.DataFrame:
+    """Gets ojo sample data with skills information from s3
+
+    Returns:
+        pd.Dataframe: ojo sample data
```
Comment (RFOxbury) [inline dap_job_quality/getters/ojo_getters.py]:
```suggestion
        pd.Dataframe: ojo sample data with the fields:
            - id: unique identifier for the job ad from which the skill was extracted
            - skill_label: the skill as extracted by our skills extractor eg "Advanced IT skills"
            - esco_label: the skill from the ESCO taxonomy that the skill maps to eg "using digital tools for collaboration and productivity"
            - esco_id: identifier for the ESCO skill in the ESCO taxonomy eg S5.6.1
```

**Review by RFOxbury — APPROVED** [review]
This looks really good! I've tried all the getters in `ojo_getters.py` and they work fine for me. I've suggested some additions to the docstrings just because I can see me coming back to this in a couple of weeks and having no idea what is in the data... Anyway you can ignore them if they're not helpful or not in line with DSP practices

Also left a question about the Python version because somehow I've ended up on 3.11... Go ahead and merge if you think that is unlikely to be a problem. I'd rather have 3.11 than 3.8 but I don't understand why it's ended up on 3.11 when 3.10 is specified in the yaml!

And finally: there's a warning about s3fs that looks more doom-laden than most warnings 😖 Maybe I just have a weird s3fs version installed? I'm on `0.4.2`

File: environment.yaml (hunk 1 above)
Comment (india-kerle) [reply environment.yaml:6]: you did run make install! this is v weird...

Comment (india-kerle) [reply environment.yaml:6]: concerned lol as i think we should all be using the same python version given potential dependencies/conflicts - guess you could just downgrade?

File: dap_job_quality/utils/bert_vectorizer.py (hunk 2 above)
Comment (india-kerle) [reply dap_job_quality/utils/bert_vectorizer.py:41]: it's old code!!

File: requirements.txt (hunk 3 above)
Comment (india-kerle) [reply requirements.txt]: i didn't get that warning with 3.10.13 python - also v weird given we don't define the version so you'd think you would install the latest release?

File: dap_job_quality/getters/ojo_getters.py (hunk 4 above)
Comment (india-kerle) [reply dap_job_quality/getters/ojo_getters.py]: great idea!

**Comment by india-kerle** [comment]
thanks for the Makefile change and good catch @RFOxbury ! I've made all your suggested changes and pinned the s3fs version so hopefully that error message doesn't come up. Will merge accordingly 💪

**Pull request merged**